src/haddock/libs/libgrid.py

```diff
@@ -0,0 +1,508 @@
+        # Expect the commands to be in the PATH
+        which_cmd = shutil.which(cmd)
+        if not which_cmd:
+            log.error(f"Command '{cmd}' not found in PATH.")
+            return False
+
+    return True
+
+
+class JobStatus(Enum):
+    WAITING = "Waiting"
+    RUNNING = "Running"
+    UNKNOWN = "Unknown"
+    DONE = "Done"
+    MATCHED = "Matched"
+    COMPLETING = "Completing"
+    FAILED = "Failed"
+
+    @classmethod
+    def from_string(cls, value):
+        """Convert string to JobStatus enum."""
+        value = value.strip().lower()
+        for status in cls:
+            if status.value.lower() == value:
+                return status
+        return cls.UNKNOWN
+
+
+class GridJob:
+    """GridJob is a class tha represents a job to be run on the GRID via DIRAC."""
+
+    def __init__(self, cnsjob: CNSJob) -> None:
+        # Unique name for the job
+        self.name = str(uuid.uuid4())
+        # Create a temporary directory for the job
+        self.loc = Path(tempfile.mkdtemp(prefix="haddock_grid_"))
+        # working directory where the GridJob was created, this is important
+        #  so we know where to put the files coming from the grid
+        self.wd = Path.cwd()
+        self.input_f = self.loc / "input.inp"
+        # `cns_out_f` is the expected `.out` file that will be generated by this execution
+        self.cns_out_f = cnsjob.output_file
+        # `expected_outputs` is what will be produced by this `.inp` file
+        self.expected_outputs = []
+        # `id` is given by DIRAC
+        self.id = None
+        # Site where the job is running
+        self.site = None
+        # Output and error files generated by the job.sh execution
+        self.stdout_f = None
+        self.stderr_f = None
+        # Script that will be executed in the grid
+        self.job_script = self.loc / "job.sh"
+        # JDL file that describes the job to DIRAC
+        self.jdl = self.loc / "job.jdl"
+        # Internal status of the job
+        self.status = JobStatus.UNKNOWN
+        # List of files to be included in the payload
+        self.payload_fnames = []
+        # Counter for retries
+        self.retries = 0
+
+        # NOTE: `cnsjob.input_file` can be a string or a path, handle the polymorfism here
+        if isinstance(cnsjob.input_file, Path):
+            self.input_f = cnsjob.input_file
+
+        elif isinstance(cnsjob.input_file, str):
+            with open(self.input_f, "w") as f:
+                f.write(cnsjob.input_file)
+
+        # CREATE JOB FILE
+        self.create_job_script()
+
+        # CREATE PAYLOAD
+        self.prepare_payload(
+            cns_exec_path=Path(cnsjob.cns_exec),
+            cns_script_path=Path(cnsjob.envvars["MODULE"]),
+            toppar_path=Path(cnsjob.envvars["TOPPAR"]),
+        )
+
+        # CREATE THE JDL
+        self.create_jdl()
+
+    def prepare_payload(
+        self, cns_exec_path: Path, cns_script_path: Path, toppar_path: Path
+    ):
```

Comment:
Missing return value
```suggestion
    ) -> None:
```